### PR TITLE
Solved some bugs that happens from async sources...

### PR DESCRIPTION
**Pull request by tounano**
for `flow.serial()`, the following code would throw an error:

``` js
pull(
  pull.values([1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20]),
  flow.async(),
  flow.serial(),
  pull.log()
)
```

Basically the problem here is by using the following code:

``` js
output[0].end, output.shift().cb
```

The error happens when reading from async sources. I believe that it happens because of messing up with object references. Something internal for node.

regarding `flow.parallel`, the following code would throw an error:

``` js
pull(
  pull.values([1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20]),
  flow.async(),
  flow.serial(),
  flow.async(),
  flow.parallel(10),
  pull.log()
)
```

Again, some problems reading from async sources. Some reference errors.

And finally, I changed the way it does error handling.

If an error bubbles up from a reader/sink, the stream would end immediately. In case an error happened while reading readable, the error would be sent to the sink, but it won't stop executing.

### CODE REVIEW

**Comment by dominictarr** [comment]
can you describe what the problem was and what you have fixed? it's hard to figure that out just looking at a diff. Also, how did you encounter this problem, what where you doing?

**Comment by tounano** [comment]
Hey,

check out my comment:
https://github.com/dominictarr/pull-flow/pull/2#issue-36832645

I explain exactly how to reissue the error, and what's the problem.

Basically, on `flow.serial` it's a problem with how node's array references work.

`flow.parallel` had the same bug + I changed something I'm not sure you'll want by design.

If you get an error while reading from upstream, it'll send it downstream. However, it will end the stream only once end===true.

If you get an error from downstream, it will end the stream.

Basically, if `flow.serial` won't end on error and pass it downstream, flow.parallel should follow the same behavior IMHO.

**Comment by dominictarr** [comment]
You are only describing the problem in abstract...
But I can't really understand what you are trying to do unless you explain it concretely
That means an actual real world example I can run through in my head.
Context is very important.

**Comment by tounano** [comment]
I understand my mistake now.
